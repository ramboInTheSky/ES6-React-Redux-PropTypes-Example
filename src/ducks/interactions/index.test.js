import reducer, {
  initialState,
  clearInteractionError,
  interactionError,
  interactionLoading,
  setActivityTypes,
  setInteraction,
  createInteraction,
  getActivityTypes,
  SET_INTERACTION,
  SET_INTERACTION_ERROR,
  SET_ACTIVITY_TYPES,
  getThirdPartiesContact,
  SET_THIRD_PARTY,
  SET_THIRD_PARTY_ERROR,
} from './';

jest.mock('../../api', () => ({
  interactionsApi: {
    getActivityTypes: withError =>
      withError
        ? Promise.reject({ response: { status: 500 } })
        : { data: [{ id: 'a', name: 'item' }] },
    createInteraction: (id, payload) => ({ [id]: payload }),
  },
  correspondenceApi: {
    getThirdPartyRecipients: withError =>
      withError
        ? Promise.reject({ response: { status: 500 } })
        : { data: [{ id: 'a', name: 'item' }] },
  },
}));

const activityTypes = [
  {
    id: 0,
    name: 'Email',
  },
  {
    id: 1,
    name: 'InPerson',
  },
  {
    id: 2,
    name: 'Letter',
  },
  {
    id: 3,
    name: 'Messaging',
  },
  {
    id: 4,
    name: 'PhoneCall',
  },
  {
    id: 5,
    name: 'Voicemail',
  },
];

describe('interactions reducer', () => {
  let dispatch;

  beforeEach(() => {
    dispatch = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should set a default state', () => {
    expect(reducer(undefined, {})).toEqual(initialState);
  });

  it('returns untouched state by default', () => {
    const result = reducer(initialState, {});
    expect(result).toEqual(initialState);
  });

  it('should set the correct state for clearInteractionError', () => {
    const result = reducer(initialState, clearInteractionError());
    expect(result).toMatchSnapshot();
  });

  it('should set the correct state for interactionError', () => {
    const payload = {
      a: 'b',
    };
    const result = reducer(initialState, interactionError(payload));
    expect(result).toMatchSnapshot();
  });

  it('should set the correct state for interactionLoading', () => {
    const result = reducer(initialState, interactionLoading());
    expect(result).toMatchSnapshot();
  });

  it('should set the correct state for setActivityTypes', () => {
    const result = reducer(initialState, setActivityTypes(activityTypes));
    expect(result).toMatchSnapshot();
  });

  it('should set the correct state for setInteraction', () => {
    const result = reducer(initialState, setInteraction({ a: 'b' }));
    expect(result).toMatchSnapshot();
  });

  it('should perform the correct actions for createInteraction', async () => {
    const arrearId = 'abd';
    const payload = { a: 'b' };
    await createInteraction(arrearId, payload)(dispatch);
    expect(dispatch).toBeCalledWith({ type: SET_INTERACTION, payload: { [arrearId]: payload } });
    expect(dispatch).toBeCalled();
  });

  it('should perform the correct actions for getActivityTypes', async () => {
    await getActivityTypes()(dispatch);
    expect(dispatch).toBeCalledWith({
      type: SET_ACTIVITY_TYPES,
      payload: [{ id: 'a', name: 'item' }],
    });
  });

  describe('getActivityTypes error', () => {
    it('should perform the correct actions for getActivityTypes in case of an error', async () => {
      await getActivityTypes(true)(dispatch).catch(() => {
        expect(dispatch).toBeCalledWith({ type: SET_INTERACTION_ERROR });
      });
    });
  });

  it('should perform the correct actions for getThirdPartiesContact', async () => {
    await getThirdPartiesContact()(dispatch);
    expect(dispatch).toBeCalledWith({
      type: SET_THIRD_PARTY,
      payload: [{ id: 'a', name: 'item' }],
    });
  });

  it('should perform the correct actions when getThirdPartiesContact fails', async () => {
    await getThirdPartiesContact(true)(dispatch);
    expect(dispatch).toBeCalledWith({
      type: SET_THIRD_PARTY_ERROR,
    });
  });
});
