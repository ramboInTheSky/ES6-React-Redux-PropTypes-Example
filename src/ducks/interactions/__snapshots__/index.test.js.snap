// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interactions reducer should set the correct state for clearInteractionError 1`] = `
Object {
  "activity": Object {},
  "activityTypes": Array [],
  "error": false,
  "loading": false,
  "thirdParties": undefined,
  "thirdPartiesError": false,
}
`;

exports[`interactions reducer should set the correct state for interactionError 1`] = `
Object {
  "activity": Object {},
  "activityTypes": Array [],
  "error": true,
  "loading": false,
  "thirdParties": undefined,
  "thirdPartiesError": false,
}
`;

exports[`interactions reducer should set the correct state for interactionLoading 1`] = `
Object {
  "activity": Object {},
  "activityTypes": Array [],
  "error": false,
  "loading": true,
  "thirdParties": undefined,
  "thirdPartiesError": false,
}
`;

exports[`interactions reducer should set the correct state for setActivityTypes 1`] = `
Object {
  "activity": Object {},
  "activityTypes": Array [
    Object {
      "id": 0,
      "name": "Email",
    },
    Object {
      "id": 1,
      "name": "In Person",
    },
    Object {
      "id": 2,
      "name": "Letter",
    },
    Object {
      "id": 3,
      "name": "Messaging",
    },
    Object {
      "id": 4,
      "name": "Phone Call",
    },
    Object {
      "id": 5,
      "name": "Voicemail",
    },
  ],
  "error": false,
  "loading": false,
  "thirdParties": undefined,
  "thirdPartiesError": false,
}
`;

exports[`interactions reducer should set the correct state for setInteraction 1`] = `
Object {
  "activity": Object {},
  "activityTypes": Array [],
  "error": false,
  "loading": false,
  "payload": Object {
    "a": "b",
  },
  "thirdParties": undefined,
  "thirdPartiesError": false,
}
`;
